File: Caster/Interface/Themes/Theme.swift
import SwiftUI

/// A colour palette. `LightTheme` and `DarkTheme` are the two concrete values;
/// views read `\.theme` from the environment rather than hard-coding either one,
/// which is what keeps the app honest in Dark Mode.
struct Theme {
    let background: Color
    let surface: Color
    let surfaceRaised: Color
    let border: Color
    let textPrimary: Color
    let textSecondary: Color
    let accent: Color
    let success: Color
    let danger: Color
    let warning: Color
    let playerColors: [Color]

    /// Colour for a seat, wrapping around when there are more players than colours.
    func playerColor(for index: Int) -> Color {
        guard !playerColors.isEmpty else { return accent }
        // `%` on a negative index would be negative and trap the subscript.
        return playerColors[abs(index) % playerColors.count]
    }

    static func forScheme(_ scheme: ColorScheme) -> Theme {
        scheme == .dark ? DarkTheme.palette : LightTheme.palette
    }
}

/// The eight seat colours, shared by both palettes.
enum PlayerPalette {
    static let colors: [Color] = [
        Color(red: 0.235, green: 0.506, blue: 0.882),  // 1 — Blue
        Color(red: 0.063, green: 0.690, blue: 0.498),  // 2 — Green
        Color(red: 0.859, green: 0.780, blue: 0.384),  // 3 — Amber
        Color(red: 0.560, green: 0.365, blue: 0.250),  // 4 — Brown
        Color(red: 0.533, green: 0.373, blue: 0.831),  // 5 — Purple
        Color(red: 0.961, green: 0.298, blue: 0.486),  // 6 — Pink
        Color(red: 0.063, green: 0.690, blue: 0.859),  // 7 — Cyan
        Color(red: 0.392, green: 0.447, blue: 0.509),  // 8 — Slate
    ]

    /// A colour for slice `index` of `total`. The pinwheel is unbounded, so it
    /// cannot use the fixed eight: past that the hue is spread evenly around the
    /// wheel instead, which keeps neighbours distinguishable at any count.
    static func spread(index: Int, total: Int) -> Color {
        guard total > colors.count else {
            return colors[abs(index) % colors.count]
        }
        let position = Double(abs(index) % max(1, total)) / Double(max(1, total))
        // Walk the hue with a large stride so adjacent slices are far apart on
        // the wheel rather than a slow gradient that reads as one blur.
        let hue = (position * 1.0 + Double(index % 3) * 0.11).truncatingRemainder(dividingBy: 1.0)
        let saturation = index.isMultiple(of: 2) ? 0.62 : 0.74
        let brightness = index.isMultiple(of: 3) ? 0.92 : 0.80
        return Color(hue: hue, saturation: saturation, brightness: brightness)
    }
}

private struct ThemeKey: EnvironmentKey {
    static let defaultValue = LightTheme.palette
}

extension EnvironmentValues {
    var theme: Theme {
        get { self[ThemeKey.self] }
        set { self[ThemeKey.self] = newValue }
    }
}

File: android/app/src/main/java/com/jesperhaafkes/caster/ui/theme/Theme.kt
package com.jesperhaafkes.caster.ui.theme

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import kotlin.math.abs

/**
 * A colour palette. [LightTheme] and [DarkTheme] are the two concrete values;
 * screens read [LocalTheme] rather than hard-coding either one, which is what
 * keeps the app honest in dark mode.
 */
data class Theme(
    val background: Color,
    val surface: Color,
    val surfaceRaised: Color,
    val border: Color,
    val textPrimary: Color,
    val textSecondary: Color,
    val accent: Color,
    val success: Color,
    val danger: Color,
    val warning: Color,
    val playerColors: List<Color>,
) {
    /** Colour for a seat, wrapping around when there are more players than colours. */
    fun playerColor(index: Int): Color {
        if (playerColors.isEmpty()) return accent
        // A negative index would otherwise walk off the front of the list.
        return playerColors[abs(index) % playerColors.size]
    }
}

val LightTheme = Theme(
    background = Color(1.0f, 1.0f, 1.0f),
    surface = Color(1.0f, 1.0f, 1.0f),
    surfaceRaised = Color(0.949f, 0.949f, 0.949f),
    border = Color(0.902f, 0.902f, 0.902f),
    textPrimary = Color(0.063f, 0.063f, 0.063f),
    textSecondary = Color(0.404f, 0.404f, 0.404f),
    accent = Color(0.235f, 0.506f, 0.882f),
    success = Color(0.047f, 0.639f, 0.427f),
    danger = Color(0.855f, 0.204f, 0.263f),
    warning = Color(0.918f, 0.639f, 0.145f),
    playerColors = PlayerPalette.colors,
)

val DarkTheme = Theme(
    background = Color(0.027f, 0.027f, 0.030f),
    surface = Color(0.047f, 0.047f, 0.047f),
    surfaceRaised = Color(0.106f, 0.106f, 0.106f),
    border = Color(0.182f, 0.182f, 0.182f),
    textPrimary = Color(0.965f, 0.965f, 0.965f),
    textSecondary = Color(0.569f, 0.569f, 0.569f),
    accent = Color(0.235f, 0.506f, 0.882f),
    success = Color(0.157f, 0.780f, 0.549f),
    danger = Color(0.937f, 0.325f, 0.376f),
    warning = Color(0.976f, 0.749f, 0.290f),
    playerColors = PlayerPalette.colors,
)

/** The eight seat colours, shared by both palettes. */
object PlayerPalette {
    val colors: List<Color> = listOf(
        Color(0.235f, 0.506f, 0.882f), // 1 — Blue
        Color(0.063f, 0.690f, 0.498f), // 2 — Green
        Color(0.859f, 0.780f, 0.384f), // 3 — Amber
        Color(0.560f, 0.365f, 0.250f), // 4 — Brown
        Color(0.533f, 0.373f, 0.831f), // 5 — Purple
        Color(0.961f, 0.298f, 0.486f), // 6 — Pink
        Color(0.063f, 0.690f, 0.859f), // 7 — Cyan
        Color(0.392f, 0.447f, 0.509f), // 8 — Slate
    )

    /**
     * A colour for slice [index] of [total]. The pinwheel is unbounded, so it
     * cannot use the fixed eight: past that the hue is spread evenly around the
     * wheel instead, which keeps neighbours distinguishable at any count.
     */
    fun spread(index: Int, total: Int): Color {
        if (total <= colors.size) return colors[abs(index) % colors.size]

        val position = (abs(index) % maxOf(1, total)).toFloat() / maxOf(1, total).toFloat()
        // Walk the hue with a large stride so adjacent slices are far apart on
        // the wheel rather than a slow gradient that reads as one blur.
        val hue = ((position + (index % 3) * 0.11f) % 1.0f) * 360f
        val saturation = if (index % 2 == 0) 0.62f else 0.74f
        val brightness = if (index % 3 == 0) 0.92f else 0.80f
        return Color.hsv(hue, saturation, brightness)
    }
}

val LocalTheme = staticCompositionLocalOf { LightTheme }

/** The palette matching the system colour scheme, tracked live. */
@Composable
@ReadOnlyComposable
fun themeForScheme(): Theme = if (isSystemInDarkTheme()) DarkTheme else LightTheme
